File: apps/research/research_pipeline/paths.py
from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path


def find_repo_root(start: Path | None = None) -> Path:
    start = (start or Path.cwd()).resolve()
    for candidate in [start, *start.parents]:
        if (candidate / "contracts" / "MODELS.json").is_file():
            return candidate
    # apps/research/research_pipeline → repo root
    module_root = Path(__file__).resolve().parents[3]
    if (module_root / "contracts" / "MODELS.json").is_file():
        return module_root
    raise RuntimeError(
        "Could not find repo root (expected contracts/MODELS.json). "
        "Run from the monorepo root."
    )


def repo_path(*parts: str) -> Path:
    return find_repo_root().joinpath(*parts)


@lru_cache(maxsize=1)
def load_models() -> dict:
    path = repo_path("contracts", "MODELS.json")
    return json.loads(path.read_text(encoding="utf-8"))

File: apps/generator/src/paths.ts
import fs from "fs";
import path from "path";

/** Walk up from cwd / this file until contracts/MODELS.json is found. */
export function findRepoRoot(startDir = process.cwd()): string {
  let dir = path.resolve(startDir);
  for (let i = 0; i < 12; i++) {
    if (fs.existsSync(path.join(dir, "contracts", "MODELS.json"))) {
      return dir;
    }
    const parent = path.dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  // Fallback: apps/generator/src/paths.ts → repo root is ../../..
  const fromModule = path.resolve(__dirname, "../../..");
  if (fs.existsSync(path.join(fromModule, "contracts", "MODELS.json"))) {
    return fromModule;
  }
  throw new Error(
    "Could not find repo root (expected contracts/MODELS.json). Run from the monorepo root."
  );
}

export function repoPath(...segments: string[]): string {
  return path.join(findRepoRoot(), ...segments);
}
